### PR TITLE
LoongArch build fixes

### PR DESCRIPTION
**Pull request by xen0n**

### CODE REVIEW

File: dev-libs/libffi/libffi-3.4.2-r1.ebuild [hunk 1]
```diff
@@ -10,7 +10,8 @@ MY_P=${PN}-${MY_PV}
 
 DESCRIPTION="a portable, high level programming interface to various calling conventions"
 HOMEPAGE="https://sourceware.org/libffi/"
-SRC_URI="https://github.com/libffi/libffi/releases/download/v${MY_PV}/${MY_P}.tar.gz"
+SRC_URI="https://github.com/libffi/libffi/releases/download/v${MY_PV}/${MY_P}.tar.gz
+	https://dev.gentoo.org/~xen0n/distfiles/${MY_P}-loongarch64-20220425.patch.xz"
```
Comment (akhuettel) [inline dev-libs/libffi/libffi-3.4.2-r1.ebuild:14]: This touches so much stuff outside loongarch, you should probably gate it with USE=experimental-loong. 
In addition, the patch touches automake source files, so you need to run eautoreconf from autotools.eclass in src_prepare.

**Review by akhuettel — CHANGES_REQUESTED** [review]

File: dev-libs/userspace-rcu/files/userspace-rcu-0.13.1-loong.patch [hunk 2]
```diff
@@ -0,0 +1,203 @@
+From 9527dea023eb06e49569da12fd286e7e7f568cbb Mon Sep 17 00:00:00 2001
```
Comment (thesamesam) [inline dev-libs/userspace-rcu/files/userspace-rcu-0.13.1-loong.patch:1]: Please always include links to where patches come from, upstream submissions, etc.

Also, run scrub-patch.

File: sys-libs/libseccomp/libseccomp-2.5.4.ebuild [hunk 3]
```diff
@@ -39,6 +40,7 @@ BDEPEND="${DEPEND}
 PATCHES=(
 	"${FILESDIR}"/libseccomp-python-shared.patch
 	"${FILESDIR}"/libseccomp-2.5.3-skip-valgrind.patch
+	"${WORKDIR}"/libseccomp-2.5.4-loongarch64-20220425.patch
```
[the remark this answers is not in the document]
Comment (thesamesam) [reply sys-libs/libseccomp/libseccomp-2.5.4.ebuild:43]: Yeah, we did the same thing (snapshot/rc approach) with riscv for libseccomp IIRC. At a skim I don't see the non-looongarch patches BUT I'm only looking at https://github.com/seccomp/libseccomp/pull/356/files, didn't look very closely, and the patch being applied might be more than that PR anyway.

Agreed with dilf

File: dev-libs/libffi/libffi-3.4.2-r1.ebuild (hunk 1 above)
Comment (thesamesam) [reply dev-libs/libffi/libffi-3.4.2-r1.ebuild:14]: 1. yes
2. libffi is early enough that we _might_ have to avoid it (eautoreconf) given python needs it

Comment (xen0n) [reply dev-libs/libffi/libffi-3.4.2-r1.ebuild:14]: Sure enough; libffi can't autoconf due to `autoconf -> m4 -> gettext -> libxml2[python] -> python -> libffi`. I'll adjust this.

File: dev-libs/userspace-rcu/userspace-rcu-0.13.1.ebuild [hunk 4]
```diff
@@ -19,6 +19,7 @@ BDEPEND="test? ( sys-process/time )"
 
 PATCHES=(
 	"${FILESDIR}"/${PN}-0.13.1-tests-no-benchmark.patch
+	"${FILESDIR}"/${PN}-0.13.1-loong.patch
```
[the remark this answers is not in the document]
Comment (xen0n) [reply dev-libs/userspace-rcu/userspace-rcu-0.13.1.ebuild:22]: Ah, sure. I keep forgetting about eautoreconf, maybe worked too much with upstream sources :facepalm:

[the remark this answers is not in the document]
Comment (xen0n) [reply dev-libs/userspace-rcu/userspace-rcu-0.13.1.ebuild:22]: Actually the `eautoreconf` is already there for ages... added (another) comment on top of that.

File: dev-libs/libffi/libffi-3.4.2-r1.ebuild (hunk 1 above)
Comment (xen0n) [reply dev-libs/libffi/libffi-3.4.2-r1.ebuild]: Done. (gating with `USE=experimental-loong` and verifying that autoconf isn't invoked during build)

File: dev-libs/userspace-rcu/files/userspace-rcu-0.13.1-loong.patch (hunk 2 above)
Comment (xen0n) [reply dev-libs/userspace-rcu/files/userspace-rcu-0.13.1-loong.patch]: Fixed.

File: sys-libs/libseccomp/libseccomp-2.5.4.ebuild (hunk 3 above)
[the remark this answers is not in the document]
Comment (xen0n) [reply sys-libs/libseccomp/libseccomp-2.5.4.ebuild]: Done; gated patching with `USE=experimental-loong` and verified that it doesn't invoke autoconf when building for loong.

File: dev-libs/userspace-rcu/files/userspace-rcu-0.13.1-loong.patch [hunk 5]
```diff
@@ -0,0 +1,177 @@
+This is https://github.com/loongarch64/userspace-rcu/pull/2, yet to be
+upstreamed but already reviewed and accepted by the LoongArch porting
+community.
+From: Wang Jing <wangjing@loongson.cn>
```
Comment (thesamesam) [inline dev-libs/userspace-rcu/files/userspace-rcu-0.13.1-loong.patch:4]: For unmerged PRs, I keep the commit hash so we know which version, but I guess the date is good enough

File: dev-libs/userspace-rcu/userspace-rcu-0.13.1.ebuild [hunk 6]
```diff
@@ -19,13 +19,15 @@ BDEPEND="test? ( sys-process/time )"
...
 )
 
 src_prepare() {
 	default
 
 	# Needed for tests patch
 	# ... and refresh libtool (see https://github.com/gentoo/gentoo/pull/23973)
+	# ... and for build on loong (see https://github.com/gentoo/gentoo/pull/25189)
```
Comment (thesamesam) [inline dev-libs/userspace-rcu/userspace-rcu-0.13.1.ebuild:30]: Thanks for adding that bit on!

**Review by thesamesam — APPROVED** [review]

File: sys-libs/libseccomp/libseccomp-2.5.4.ebuild [hunk 7]
```diff
@@ -42,8 +43,19 @@ PATCHES=(
 )
 
 src_prepare() {
+	if use experimental-loong; then
+		PATCHES+=( "${WORKDIR}"/libseccomp-2.5.4-loongarch64-20220425.patch )
```
Comment (thesamesam) [inline sys-libs/libseccomp/libseccomp-2.5.4.ebuild:47]: use P here to be consistent with SRC_URI

Comment (xen0n) [reply sys-libs/libseccomp/libseccomp-2.5.4.ebuild]: done

**Review by akhuettel — APPROVED** [review]

**Review by thesamesam — APPROVED** [review]

**Pull request closed without merging**